lib/src/debian.dart

```diff
@@ -0,0 +1,216 @@
+  _removeDirectory(debianDir);
+}
+
+/// Delete the Directory with path [directory].
+Future<void> _removeDirectory(String directory) async {
+  var result = await Process.run("rm", ["-r", directory]);
+  if (result.exitCode != 0) {
+    fail('Unable to remove the directory\n${result.stderr}');
+  }
+}
+
+/// Scans the PPA [repo] for new packages and updates the
+/// release files, also signing them.
+Future<void> _releaseNewPackage(String repo) async {
+  await _updatePackagesFile(repo);
+  await _updateReleaseFile(repo);
+  await _updateReleaseGPGFile(repo);
+  await _updateInReleaseFile(repo);
+}
+
+/// Create the directory `repo/packageName` and relevant subfolders for the
+/// debian package.
+///
+/// Returns the path of created folder.
+Future<String> _createPackageDirectory(String repo, String packageName) async {
+  String debianDir = p.join(repo, packageName);
+  await Directory('$debianDir/DEBIAN').create(recursive: true);
+  await Directory('$debianDir/usr/local/bin').create(recursive: true);
+  return debianDir;
+}
+
+/// Generate the executable files from the map [executables] which contains the
+/// executable name as key and the path to the corresponding Dart file as value.
+void _generateExecutableFiles(String debianDir) {
+  final executablePath = p.join(debianDir, "usr", "local", "bin");
+  executables.value.forEach((name, path) {
+    run('dart', arguments: [
+      'compile',
+      'exe',
+      path,
+      for (var entry in environmentConstants.value.entries)
+        '-D${entry.key}=${entry.value}',
+      '--output',
+      p.join(executablePath, name)
+    ]);
+  });
+}
+
+/// Generate the control file for the Debian package.
+void _generateControlFile(String debianDir) {
+  var controlFilePath = p.join(debianDir, "DEBIAN", "control");
+  writeString(controlFilePath, controlData.value);
+}
+
+/// Scan for new .deb packages in the [repo] and update the `Packages` file.
+Future<void> _updatePackagesFile(String repo) async {
+  // Scan for new packages
+  String output = run("dpkg-scanpackages",
+      arguments: ["--multiversion", "."], workingDirectory: repo);
+  // Write the stdout to the file
+  writeString(p.join(repo, 'Packages'), output);
+  // Force Compress the Packages file
+  run("gzip", arguments: ["-k", "-f", "Packages"], workingDirectory: repo);
+}
+
+/// Generate the Release index for the PPA.
+Future<void> _updateReleaseFile(String repo) async {
+  String output = run("apt-ftparchive",
+      arguments: ["release", "."], workingDirectory: repo);
+  writeString(p.join(repo, 'Release'), output);
+}
+
+/// Sign the Release file with the GPG key.
+Future<void> _updateReleaseGPGFile(String repo) async {
+  String output = run("gpg",
+      arguments: [
+        "--default-key",
+        signingEmail.value,
+        "-abs",
+        "-o",
+        "-",
+        "Release",
+      ],
+      workingDirectory: repo);
+  writeString(p.join(repo, 'Release.gpg'), output);
+}
```

Comment:
I don't completely understand how this signing process works. Is there any verification that the package author actually owns the email in question? Does it use a private key at all? And if not, what does the signature actually prove?

Reply:
GPG keys are tied to an email address.

`apt` verifies the signature of APT repositories. So the public key will need to be published, and getting it is part of the instructions to add the apt repository (see for instance the Linux instructions about [installing the Dart SDK](https://dart.dev/get-dart) which first download the public key and pass it to `apt-key add` and then add the APT repository).
The private GPG key is used by this `gpg` command to perform the signing

Reply:
Also, we can export the private and public keys from `gpg` to use on a remote computer (GitHub Actions, in our case) using:

```sh
gpg --output mygpgkey_pub.gpg --armor --export <keyId> # For public key
gpg --output mygpgkey_sec.gpg --armor --export-secret-key <keyId> # For private key
```

And import these to `gpg` using:

```sh
gpg --import ./mygpgkey_pub.gpg
gpg --allow-secret-key-import --import ./mygpgkey_sec.gpg
```

Ref: https://www.debuntu.org/how-to-importexport-gpg-key-pair/

Reply:
There is https://github.com/marketplace/actions/import-gpg implementing the import of GPG keys for github actions (with proper cleanup to remove the key at the end of the workflow run)

Reply:
So if I'm understanding correctly, GPG has some sort of database on the system that it queries using the email as a key? That's a lot to ask an unfamiliar user to set up correctly. Can we provide an optional `aptPrivateKey` config variable that allows users to directly set the private key as a GH Actions secret?

Reply:
@nex3 I'd rather rely on the dedicated action to import the GPG key (which only requires setting the GPG private key and its passphrase as secret and exposes some output based on the extracted id), as that action also hooks into the cleanup phase of GA to remove the imported key again, avoiding to leak it to the runner.

Reply:
The workflow would look like that:

```yaml
    steps:
      -
        name: Checkout
        uses: actions/checkout@v2
      -
        name: Import GPG key
        id: import_gpg
        uses: crazy-max/ghaction-import-gpg@v4
        with:
          gpg_private_key: ${{ secrets.GPG_PRIVATE_KEY }}
          passphrase: ${{ secrets.PASSPHRASE }}
      -
        name: Release APT
        run: dart run grinder publish-apt
        env:
          GPG_SIGNING_KEY: ${{ steps.import_gpg.outputs.fingerprint }}
```

Reply:
If I understood correctly, the main reason for using the actions here is to avoid leakage of the key? @stof

Alternatively, we can import the key from the private key and passphrase, sign the release, and cleanup keys, right from this package. But that's essentially what the action already does :thinking:.

Reply:
I vote for letting the action handle all that, which is less code to maintain (and makes the dart_cli_pkg implementation compatible with systems where the GPG key management would have different needs due to the way the GPG key itself is managed)

Reply:
note however that I'm not a maintainer here (or in sass), so @nex3 is the one having the final call.

Reply:
Sorry, I wasn't well for the last few days so wasn't quite active on GitHub. I am on the neutral side here.

##### Reason:
- Less maintenance and no worries about cleanup (pro)
- An extra step that needs to be added manually (con)

What do you say, @nex3?

Reply:
I'd like to support the full signing workflow directly within this package. We can still provide the ability to get the key from the ambient GPG context, which can then be configured using `crazy-max/ghaction-import-gpg`, but one of the goals of this package is to be "batteries-included" and require as little setup as possible.

Reply:
Well, the thing is that the current setup still relies on the ambient GPG context (AFAIK, it is not possible to run a self-contained GPG signing command that provides the key directly instead of using the trust store). So this would still depend on having the GPG command available and modifying that ambient context.

Reply:
If I understood correctly, @nex3, do you mean to say the package shall support two modes as follows?
- use the key from context, i.e., already present in the system. This can also be used if we run the package locally or on a computer where the key is already loaded. Or even in this case, the GitHub action was used to configure GPG.
- load key and then sign using it. (pretty much the release can be made without any external steps)

And @stof, you are saying - even if we code for importing the key, we'll still be using the same GPG commands and do the same stuff like that in the GitHub action?

Reply:
> Well, the thing is that the current setup still relies on the ambient GPG context (AFAIK, it is not possible to run a self-contained GPG signing command that provides the key directly instead of using the trust store). So this would still depend on having the GPG command available and modifying that ambient context.

It's not modifying the ambient context that bothers me, it's the DevX overhead of setting it up.

> If I understood correctly, @nex3, do you mean to say the package shall support two modes as follows?
> 
> * use the key from context, i.e., already present in the system. This can also be used if we run the package locally or on a computer where the key is already loaded. Or even in this case, the GitHub action was used to configure GPG.
> * load key and then sign using it. (pretty much the release can be made without any external steps)

That's right. Basically, we can provide a `pkg.gpgPrivateKey` config variable and, if that's not set, fall back on `pkg.gpgFingerprint` or `pkg.gpgEmail` to look up an existing key from the store.